Introduce version file support into develop branch

KateFriedman-NOAA opened the pull request:
**Description**

This PR introduces the use of version files in the `develop` branch. This initial introduction will use version files for runtime only. The ability to force version files for build is included but commented out for now. A follow-up PR will turn on build version forcing when ready.

Tagging @DavidHuber-NOAA about Jet/S4 updates.

Resolves #671
Resolves #1625 

Changes include:

1. Update `module_base.*.lua` modulefiles to use module version variables and remove hardcoded versions.
2. Introduce `build.<platform>.ver`, `run.<platform>.ver` files in the `/versions` folder.
3. Update `sorc/link_workflow.sh` to create `build.ver` and `run.ver` under `/versions` folder based on `build.<platform>.ver` and `run.<platform>.ver`.
4. Introduce sourcing of `run.ver` in `ush/load_fv3gfs_modules.sh`
5. Introduce sourcing of `build.ver` in `sorc/build_all.sh` but commented out for now. Components will continue setting their module versions for the time being.
6. Update `HOMEens_tracker` in `config.vrfy` to use `tracker_ver` from version files.
7. Replace obsproc version in `HOMEobsproc` setting in `parm/config/gfs/config.base.emc.dyn` to use `obsproc_run_ver` from version files.

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

- [x] Clone and Build tests on Hera, Orion, WCOSS2 (no build changes though)
- [x] Cycled tests on Hera, Orion, WCOSS2 (1.5 cycles - reproduced control with `develop`)
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

KateFriedman-NOAA committed:
Move transfer files into new transfer folder

- In move to WCOSS2 the transfer*list files were moved into a new
transfer folder under the upper-level parm folder.

Refs: #419

KateFriedman-NOAA committed:
Updated transfer list files from WCOSS2 ops

- The transfer*list files were updated in operations during the move
from WCOSS1 to WCOSS2.
- The beginning of the path within the transfer*list files was updated
from:
  com/gfs/_ENVIR_
to:
  _COMROOT_/gfs/_SHORTVER_

Refs: #419

KateFriedman-NOAA committed:
Add GFSv16.1 and GFSv16.2 release notes

- Add release notes for GFSv16.1 and GFSv16.2 versions that were added
into the operations branch but not yet in the develop branch.

Refs: #419

KateFriedman-NOAA committed:
Merge branch 'develop' into feature/dev-wcoss2

* develop:
  Updated GFS transfer*list files from operations (develop) (#918)

KateFriedman-NOAA committed:
WCOSS2 updates to gempak ush scripts

- Add /gempak subfolder where needed in gempak ush scripts.
- Remove unneeded commented out path settings from older iterations.

Refs: #419

KateFriedman-NOAA committed:
Removing older release notes

- Cleaning out older GFS version release notes.
- Includes current GFSv16.2.1 release notes, will commit GFSv16.3
release notes with implementation this fall.
- Will then keep only the latest release notes moving forward.

Refs: #419

KateFriedman-NOAA committed:
Merge branch 'develop' into feature/dev-wcoss2

* develop:
  Add postsnd job when bufrsnd it on (#926)

KateFriedman-NOAA committed:
Add versions folder and version files

- New folder called "versions" for holding .ver files
- New build.ver for setting versions at buildtime
- New run.ver for setting versions at runtime
- New $target.ver files for setting additional non-ops developer
versions for support target platforms (WCOSS2, Hera, Orion)

Refs: #419

KateFriedman-NOAA committed:
Add LUA module_base for WCOSS2

Refs: #419

KateFriedman-NOAA committed:
Merge branch 'develop' into feature/dev-wcoss2

* develop:
  Add GDASapp (first wave of JEDI changes) (#871)

KateFriedman-NOAA committed:
Merge branch 'develop' into feature/dev-wcoss2

* develop:
  WCOSS2 gempak ush scripts updates and cleanup of old release notes (#920)

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:feature/dev-wcoss2' into feature/dev-wcoss2

KateFriedman-NOAA committed:
Merge branch 'develop' into feature/dev-wcoss2

* develop:
  change gdasechgres dependency to just gdasefcs01 instead of gdasefmn (#933)
  Remove obsolete platforms (WCOSS1, Dell, Cray, Theia) references. (#922)
  Optimize DA clone, build, and link (#931)

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:feature/dev-wcoss2' into feature/dev-wcoss2

KateFriedman-NOAA committed:
Add WCOSS2 to link_workflow.sh

Refs: #419

KateFriedman-NOAA committed:
Add WCOSS2 machine-setup.sh and add module reset

- Add WCOSS2 block to machine-setup.sh
- Update the Hera and Orion blocks to use module reset instead of module
purge.
- Add version file sourcing (build.ver and $target.ver).

Refs: #419

KateFriedman-NOAA committed:
Add WCOSS2 to module-setup.sh.inc

Refs: #419

KateFriedman-NOAA committed:
Create WCOSS2 build modulefiles

- Create modulefile.ww3.wcoss2.lua and workflow_utils.wcoss2.lua with
version variables within for building WW3 and workflow_utils.

Refs: #419

KateFriedman-NOAA committed:
Introduce version variables into modulefiles

- Replace hardcoded module versions with version variables in runtime
and build modulefiles for Hera, Orion, and Jet.
- Module versions will be sources from new version files at build or
runtime.

Refs: #671

KateFriedman-NOAA committed:
Merge branch 'develop' into feature/dev-wcoss2

* develop:
  Initial commit of directory comparison tools (#934)
  Add preamble, convert to bash, and remove env (#929)
  minimal intervention to create a data-atmosphere xml (#936)
  Remove Cray, Dell, WCOSS1 from module-setup.sh.inc (#943)
  bring GDASApp jjobs and exscripts to global-workflow (#941)

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:feature/dev-wcoss2' into feature/dev-wcoss2

KateFriedman-NOAA committed:
Version file updates

- Version updates within build.ver, run.ver and target specific ver
files for Hera and Orion.

Refs: #671, #419

KateFriedman-NOAA committed:
Add WCOSS2 and version files to module load script

- Add version file sourcing to load_fv3gfs_module.sh for use at runtime.
- Add WCOSS2 block to load_fv3gfs_module.sh.

Refs: #419, #671

KateFriedman-NOAA committed:
Merge branch 'develop' into feature/dev-wcoss2

* develop:
  Remove SDATE=CDATE IAU block in NCO config.base (#963)
  Updates for P8  (#946)
  Fix GLDAS j-job link (#954)
  Add ocean post to archive dependencies (#949)

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:feature/dev-wcoss2' into feature/dev-wcoss2

KateFriedman-NOAA committed:
Adjustments to hera.ver

- Updated cmake_ver to 3.20.1
- Remove esmf_ver and cdo_ver; build.ver versions good on Hera

Refs: #671

KateFriedman-NOAA committed:
Introduce version variables into workflow_utils.jet.lua

- Convert workflow_utils.jet.lua to use version variables instead of hardcoded versions

Refs: #671

KateFriedman-NOAA committed:
Change png to libpng in modulefiles

- Change references to "png" to "libpng" in modulefiles.
- If using 2022 intel modules then it's libpng.

Refs: #671

KateFriedman-NOAA committed:
Create WW3 modulefile for Jet

Refs: #671

KateFriedman-NOAA committed:
Add w3nco module back into modulefiles and version files

- Build errors without loading w3nco, will reevaluate removing
the w3nco module load later.

Refs: #671

KateFriedman-NOAA committed:
Merge branch 'develop' into feature/version-files

* develop:
  Correct issue in linking final restart files (#1285)
  Remove execute permissions from config files (#1281)
  Make needed updates to run forecast from GEFS (#1203)
  Remove unnecessary variables which reference to nemsio (#1259)
  Create analysis files for early-cycle EnKF by default (#1237)
  Don't wipe $DATA before running ocean bmat (#1280)
  More marine DA j-jobs (#1270)
  Update UFS-DA atmospheric prep script to be consistent with GDASApp update (#1265)
  Add new jjob for ocean analysis bmat (#1239)
  Retire ecf/versions in develop (#1267)
  Deploy documentation to RTD (#1264)
  Temporarily disable failing pytest (#1263)
  Remove incorrect/misleading comments in config.base (#1261)
  Add initial Sphinx documentation (#1258)
  Remove nemsio support (#1255)
  Increase wallclock for diag jobs (#1216)
  Use correct resources for GFS gempak (#1214)
  Abstract common j-job tasks (#1230)
  Add missing mkgfsawps.x link (#1218)
  Fix post sounding job (#1212)
  Revert "Use fracoro data for all new UFS applications (#1182)" (#1240)
  Use fracoro data for all new UFS applications (#1182)
  Revert "Merge GFS v16.3 operational GSI changes into develop branch. (#1158)" (#1238)
  Add more user defined parameters for the marine DA (#1235)
  Update pytests action version and run sequentially (#1236)
  Add utility to compare Fortran namelists (#1234)
  Updates for pygw (#1231)
  Merge GFS v16.3 operational GSI changes into develop branch. (#1158)
  Move member up in directory hierarchy (#1201)
  Enable staging ics for cycled experiments. (#1199)
  Add tests for configuration.py (#1192)
  Replace ocnanal_${CDATE}} with ${RUN}ocnanal_${cyc} (#1191)
  define NET and RUN in the Rocoto XML to accurately mimic the ecf in ecflow (#1193)
  Fix checking for restart files (#1186)
  Fix 'DEBUG' option in build_ufs.sh (#1188)
  Update archive job memory request value for R&Ds (#1183)
  Reorder post so all flux files are generated when running offline (#1181)
  Stop checking for restarts on non-GFS CDUMPs (#1179)
  Add missing jobids in some pre-job scripts (#1176)
  Remove existing directory if it exists when getic runs (#1165)
  Add logging decorator, test and test for yaml_file (#1178)
  fix coding norm check in `hosts.py` (#1174)
  Fix some bugs and make other changes so ctest in GDASApp works (#1172)
  Support for the GDASApp testing in containers (#1151)
  ATM 3DVAR with and without IAU (#1113)
  Enable checking for python norms and fix violating code (#1168)
  Enforce decimal math in atmos post (#1171)
  Update marine DA j-jobs to new format (#1149)
  Add utility to manipulate files en masse  (#1166)
  add action to run pytests (#1167)
  Pin `differential-shellcheck` to `v3` tag (#1162)
  Add a task base class and basic logger (#1160)
  Recursively convert dict to AttrDict when making an AttrDict (#1154)
  move configuration.py to pygw. Use it from there.  return AttrDict after sourcing configs (#1153)
  JEDI based Marine DA tasks (#1134)
  Allow customizations based on user/configuration (#1146)
  First step towards making j-jobs consistent in use from ecflow and rocoto (#1120)
  enable APP=S2SWA on WCOSS2 (#1142)
  Fix typo in .shellcheckrc
  Remove prod_envir module load from WCOSS2 (#1138)
  Link staged GSI fix files instead of cloning them from gerrit (#1132)
  Address shellcheck warnings in env files (#1136)
  Adds group size and nmem for GEFS (#1127)
  Remove unnecessary sCDATE assignment in forecast_predet.sh (#1133)
  Convert archive jobs to proper j-jobs (#1115)
  Update C48 forecast to run with one thread (#1131)
  Improved error messages from atmos analysis (#1125)
  Update MODULEPATH for Orion (#1126)
  MPMD variable updates and fix (#1124)
  Introduce FHMAX_ENKF_GFS to extending ensemble forecast capabilities (#1122)
  Update R&D launcher commands for tasks and multi-prog (#1112)
  Correct crtm path in UFS DA atmospheric analysis scripts (#1111)
  Correct syntax in remaining sorc scripts (#1105)
  Add GSI background error covariance as an option for UFS DA variational assimilation (#1104)
  Add Early Cycle EnKF workflow (#1022)
  Correct errors with gdas and monitoring symlinks (#1101)
  Fixed gfs-utils links (#1099)
  Fix build scripts and bring into compliance (#1096)
  Feature/updates for gdas app (#1091)
  Change GLDAS USE_CFP to NO on Hera (#1094)
  Resource updates to support WCOSS2 (#1070)
  Set COMPILER in link for detect machine (#1092)
  gfs utils update (#1088)
  GFS-UTILS update for build and ush scripts (#1082)
  Update UFS version to 2022 Oct 19 (#1083)
  Use more cycledefs for task control (#1078)
  removing superfluous EFSOI-specific files from develop (#1079)
  Update UFS to Sept 9 version (#1073)
  Modify default file location for monitor data when using rocoto (#1065)
  Fix companion ocean resolution for C48 (#1066)
  Add trailing slash for gldas topo path (#1064)
  Limit number of CPU for post (#1061)
  Fix eupd trace (#1057)
  Port to S4 (#1023)
  Update to obsproc.v1.0.2 and prepobs.v1.0.1 (#1049)
  Add GDAS to the partial build list (#1050)
  Fix group number being treated as octal in gdas arch (#1053)
  Remove trace from link script (#1046)
  Update gfs-utils hash to 3a609ea (#1048)
  Fix link script usage statement (#1045)
  Replace preamble variable commands with functions (#1012)
  Implement fix reorg and remove gfs-utils code (#1009)
  Rename post scripts (#1038)
  Fix missing @ symbol with COMINsyn in config.base (#1039)
  WCOSS2 run support and script/config updates (#1030)
  Remove base_svn from Hera and Orion hosts files (#1036)
  initial commit for incoming yaml work (#1029)
  Fix radiance verification failing to find diag files (#1031)
  Supported resolutions on platforms and defaults for mode (#1026)
  Add GLDAS scripts & fix GLDAS job (#1018)
  Update GSI Monitor for radmon fix
  Correct shell linter config (#1013)
  Correct diagnostic file handling in ush/ozn_xtrct.sh (#1016)
  Add shell linter Github action for pull requests (#1007)
  Build updates for WCOSS2 (#1002)
  Update UFS_UTILS tag to `ufs_utils_1_8_0` (#1001)
  Fix preamble id (#996)
  Add missing "atmos" into job dependencies (#998)
  Bugfix in arch.sh to remove hardwired "htar" (#992)
  Add in stubs for aerosol DA tasks + bugfix for setup_expt where cycled and ATMA are used (#990)
  Add GSI monitor scripts (#969)
  Fix product generation at some fcst hrs (#988)
  Add initial config files for global aerosol DA (#986)
  Update diag table to remove wav-ocn coupling fields (#979)
  use a robust Findwgrib2.cmake to find wgrib2 built w/ native wgrib2 build (#970)
  Externals.cfg was stale and had drifted off (#965)
  Fix post comparison with zero-padded numbers (#964)

KateFriedman-NOAA committed:
Removing modulefiles that were deleted in develop

Refs #671

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge remote-tracking branch 'origin/develop' into feature/version-files

* origin/develop:
  Add options while setting up Rocoto XML that are useful for CI (#1365)
  Revert "Add options while setting up Rocoto XML that are useful for CI (#1363)" (#1364)
  MOM6 backgrounds in cycled DA mode were hardcoded for IAU  (#1355)
  Add options while setting up Rocoto XML that are useful for CI (#1363)
  Run executables or scripts from within python. (#1341)
  Remove extra 'enkf' from folder/file names in archive scripts (#1360)
  Fix bugs to create analysis files for early cycle (#1343)
  Make necessary bugfixes to get aerosol cycling going (#1349)
  Create production_update.md (#1348)
  Update feature_request.md
  Create a template for requesting new features in the global workflow to separate from general issues (#1336)
  allow script to access eva yaml generators (#1331)
  Add license and status badges for the CI (#1332)
  Update ufs_utils hash that supports global_cycle with NoahMP. (#1315)
  Save yaml for ocn analysis post and fix bug to save logs (#1318)
  Update crtm to v2.4.0 (#1319)
  Fix python style errors (#1330)
  Fix workflow generation post times for early cycle (#1329)
  Change RUN and CDUMP for ensemble jobs (#1309)
  Provide default for DEBUG_WORKFLOW in load_ufsda_modules.sh
  Allow increments to be added for ocean and ice cycling. (#1308)
  Module base file reorder and reduction (#1306)
  Update S4 environment and module files (#1303)
  Add in initial 3DVar aerosol DA cycling capability (#1106)
  Fix name of MOM6 n restarts for future cycle points (#1307)

KateFriedman-NOAA committed:
Cleanup in Jet and Orion module_base modulefiles

Refs #671

KateFriedman-NOAA committed:
Cleanup in Hera module_base modulefile

Refs #671

KateFriedman-NOAA committed:
Add Jet and S4 version file sourcing to load_fv3gfs_module.sh

Refs #671

KateFriedman-NOAA committed:
Update obsproc and newm version in ver files

Refs #671

KateFriedman-NOAA committed:
Merge remote-tracking branch 'origin/develop' into feature/version-files

* origin/develop:
  Update to HR1 (#1197)

Refs #671

KateFriedman-NOAA committed:
Add missing double quotes around hpss module in Jet modulefile

Refs #671

KateFriedman-NOAA committed:
General version updates and reorganization

- Move version variables to better align with recent change to
reorder module loads into groups.
- Update versions to better align with current component versions.

Refs #671

KateFriedman-NOAA committed:
Add build version file sourcing to build_all.sh

Refs #671

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge remote-tracking branch 'origin/develop' into feature/version-files

* origin/develop:
  Initial blocks in place for forecast refactor work (#1466)
  Update buoys file and fix boundary point jobs (#1465)
  Move Fit2Obs to stand-alone job (#1456)
  Rework arch job dependencies (#1455)
  Stage bias files for UFSDA aerosols (#1370)
  Add logging level to logger (#1442)
  Update hash for GDASApp to db2f998 (#1443)
  Remove para module paths for ncdiag on WCOSS2 (#1437)
  Use P8 settings for C384 atm by default (#1440)
  Enforce rstprod on relevant tarballs (#1436)

KateFriedman-NOAA committed:
Merge remote-tracking branch 'origin/develop' into feature/version-files

* origin/develop:
  Fix RTD python requirements path
  Clean up RTD python install settings
  Use RTD python system packages
  Fix python RTD parameter name
  Update RTD python install
  Restore RTD python version and add requirements
  Revert python version for RTD (#1598)
  Move RTD config to root of repo (#1597)
  Add RTD config (#1596)
  Add `schema` library to manage schema for variety of input configurations. (#1567)
  Add S2SA to the allowed app list in setup expt (#1591)
  Avoid parsing group name when checking RadMon diagnostic files (#1559)
  Remove gldas from global-workflow (#1590)
  Add GEFS capability to setup_expt (#1582)
  Redo v16.3 GSI script updates to scripts/exglobal_atmos_analysis.sh (#1535)
  Fix bugs in the COM refactor of marine DA (#1566)
  Retire `getic.sh` and `init.sh` jobs from global-workflow (#1578)
  New UFS_UTILS hash for gdas_init COM reorg updates (#1581)
  Update documentation for PR standards (#1573)
  Update prepobs install copy for atm log filename fix (#1571)
  Update LandDA related job post COM refactor plus other minor fixes (#1564)
  Replace `CDATE` with `PDYcyc` combo (#1561)
  Update init.rst cycled coupled IC info (#1562)
  Use correct number of tasks in CICE when using threads (#1558)
  Remove gdas bump fix files (#1553)
  Remove remnant WAVE_RUN from archive (#1556)
  Update UFS to develop as of 2023 Apr 17 (#1509)
  Adapt the marine DA to the new COM structure (#1554)
  Update aerosol DA to use new COM structure (#1551)
  Update UFSDA ATM ens for new COM directory structure (#1538)
  Run an ensemble forecast of the coupled model (#1545)
  Fix ocean anl path in staging (#1544)
  Two minor bugfixes (#1542)
  Move guts of ocean analysis post out of j-job (#1539)
  Update UFSDA ATM anl for new COM directory structure (#1537)
  Fix incorrect ocean history tmpl during workflow generation (#1533)
  Fix ocean stating from flat structure (#1531)
  Fix warm-start IC staging (#1529)
  Quiet generate_com (#1526)
  Add ocnanlvrfy job and bugfixes. (#1514)
  Reorganize COM and refactor to use templates (#1421)
  Move GDASApp hash to stable version (#1508)
  update GDASApp hash to d34f616 (#1505) (#1506)
  ignore archiving sfluxgrbf00[124578] files when they are not present (#1498)
  Add initial land DA cycling scripts (#1351)
  Add CI cron jobs (#1476)
  Refactor UFS-DA ATM ens component to use python g-w (#1373)
  Added Fit2Obs to S4. #1489 (#1497)
  Add new task to post-process marine DA (#1485)
  Create fix file issue template (#1495)
  Update TC_tracker version to v1.1.15.6
  Update GSI-Monitor hash to reflect recent assimilation changes.
  Fit2Obs updates for package reorganization and invocation via module (#1484)
  Refactor UFS-DA ATM var component to use python g-w (#1372)
  Port the global workflow to Jet (#1301)
  Consolidate wave parm files (#1477)
  Split MPMD stdout into tasks on slurm (#1469)
  GFS fixed-file YAMLs. (#1471)

KateFriedman-NOAA committed:
Update Fit2Obs version and variable name

- Rename fit_ver to fit2obs_ver.
- Introduce fit2obs_ver into module_base files.

Refs #671

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Update tracker version

- Update tracker_ver=feature-GFSv17_com_reorg
- Rename ens_tracker_ver to tracker_ver in config.vrfy

Refs #671

KateFriedman-NOAA committed:
Create jet.ver and update R&D versions

- Add jet.ver to support RDHPCS Jet
- Add ncl_ver to version files for RDHPCS
- Remove wrf_io_ver from version files
- Adjust order of some version variables

Refs #671

KateFriedman-NOAA committed:
Add prepobs MODULEPATH to module_base.jet.lua

- Commented out for now until official version installed

Refs #671

KateFriedman-NOAA committed:
Create s4.ver

Refs #671

KateFriedman-NOAA committed:
Update version and gfs_ver to 16.3.7

Refs #671

KateFriedman-NOAA committed:
Remove double quotes from around fit2obs_ver version

Refs #671

KateFriedman-NOAA committed:
Move prod_util_ver down in hera.ver order

Refs #671

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA committed:
Disable invoking build.ver for now

- Comment out lines in build_all.sh for invoking build.ver.
- Will uncomment when we are ready to force build versions
down into components. Likely after move to EPIC stack complete.

Refs #671

KateFriedman-NOAA committed:
Update wgrib2 version from 2.0.7 to 2.0.8

- Using wgrib2 v2.0.8 on R&Ds, will also now use on WCOSS2.

Refs #671 #1625

KateFriedman-NOAA committed:
Role intel versions back to 2018

- Will update to intel 2022 with EPIC stack changes.

Refs #671

KateFriedman-NOAA committed:
Add grib_util_ver to orion.ver

- The v1.2.3 grib_util isn't available on Orion so use v1.2.2

Refs #671

KateFriedman-NOAA committed:
Change g2tmpl to 1.10.0 in build.ver

- Align version with runtime version

Refs #671

KateFriedman-NOAA committed:
Merge branch 'feature/version-files' of https://github.com/KateFriedman-NOAA/global-workflow into feature/version-files

* 'feature/version-files' of https://github.com/KateFriedman-NOAA/global-workflow:
  Add grib_util_ver to orion.ver

KateFriedman-NOAA committed:
Add missing versions to Jet and S4 version files

Refs #671

KateFriedman-NOAA committed:
Move ncdiag version up to match load order

Refs #671

KateFriedman-NOAA committed:
Move g2tmpl version up to match load order

Refs #671

KateFriedman-NOAA committed:
Add/remove versions to hera.ver

- Remove anaconda_ver (may add back in for build support?)
- Add cdo_ver and grib_util_ver

Refs #671

KateFriedman-NOAA committed:
Add libpng_ver, jasper_ver, and zlib_ver to version files

Refs #671

KateFriedman-NOAA committed:
Add curly braces around HOMEgfs variable in load_fv3gfs_modules.sh

Refs #671

KateFriedman-NOAA committed:
Replace obsproc version with variable in config.base.emc.dyn

Refs #671

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

aerorahul commented on ush/load_fv3gfs_modules.sh:
```diff
@@ -12,20 +12,27 @@ ulimit_s=$( ulimit -S -s )
 
 if [[ -d /lfs/f1 ]]; then
   # We are on WCOSS2 (Cactus or Dogwood)
+  source "${HOMEgfs}/versions/wcoss2.ver"
```
I think we should avoid a subset of versions in the `<platform>.ver`.
Instead, we should create `build.<platform>.ver` and `run.<platform>.ver` and copy as `build.ver` and `run.ver` in `link_workflow.sh` script.

A developer/user should be able to see in one file all the versions of the modules used on that machine rather than have it distributed.

KateFriedman-NOAA replied:
Sounds good...working on making this change...

KateFriedman-NOAA replied:
See updated PR changes. The new `build.<platform>.ver` files will need to be double checked ahead of turning on their usage.

KateFriedman-NOAA committed:
Rename version files

- Rename machine-specific version files to run.MACHINE.ver
- Rename build.ver to build.wcoss2.ver ahead of making new
build.MACHINE.ver files

Refs #671

KateFriedman-NOAA committed:
Add anaconda_ver into module_base.jet.lua

Refs #671

KateFriedman-NOAA committed:
Remove wcoss2.ver (now run.wcoss2.ver)

Refs #671

KateFriedman-NOAA committed:
Expand new run.MACHINE.ver files with all versions

Refs #671

KateFriedman-NOAA committed:
Create build.MACHINE.ver files

Refs #671

KateFriedman-NOAA committed:
Add feature to copy build[run].MACHINE.ver to build[run].ver

- Update sorc/link_workflow.sh to copy the machine-specific copy
of build.MACHINE.ver and run.MACHINE.ver into the versions
folder as build.ver and run.ver
- Add the copied build.ver and run.ver to .gitignore
- Remove sourcing of MACHINE.ver in ush/load_fv3gfs_modules.sh;
now just source run.ver copied by sorc/link_workflow.sh

Refs #671

KateFriedman-NOAA committed:
Merge branch 'feature/version-files' of https://github.com/KateFriedman-NOAA/global-workflow into feature/version-files

* 'feature/version-files' of https://github.com/KateFriedman-NOAA/global-workflow:
  Remove WAFS references from global-workflow (#1642)

KateFriedman-NOAA committed:
Remove sourcing of ${MACHINE_ID}.ver in build_all.sh

Refs #671

KateFriedman-NOAA commented:
Main PR comment updated after recent commits to PR branch. Running another cycle on Hera to check changes work as intended.

aerorahul reviewed (approved):
looks great.

DavidHuber-NOAA commented on versions/build.s4.ver:
```diff
@@ -0,0 +1,33 @@
+export hpc_intel_ver=18.0.4
+export hpc_impi_ver=18.0.4
+
+export cmake_ver=3.20.1
```
There isn't a module for cmake on S4.  The native version is 3.20.5 if it is desired to keep version numbers..

KateFriedman-NOAA replied:
Changed to `3.20.5` for now (done @ f1322477). All build.*.ver files will need full reviews ahead of them being used later. Thanks Dave!

DavidHuber-NOAA commented on versions/build.s4.ver:
```diff
@@ -0,0 +1,33 @@
+export hpc_ver=1.1.0
```
`hpc_ver` should be incremented to 1.2.0.  That's my bad for not doing this sooner.  Would you mind bumping it up?

KateFriedman-NOAA replied:
Updated @ f1322477.

DavidHuber-NOAA reviewed (commented):
A couple of minor changes for S4, but otherwise it looks good and loads well on both Jet and S4.

WalterKolczynski-NOAA commented:
I retriggered Orion CI because I was rebuilding the custom pyenv and thought that might have been the issue, but now Hera has failed as well.

WalterKolczynski-NOAA commented:
> Automated global-workflow Testing Results:
> 
> ```
> Machine: Orion
> Start: Wed May 24 18:06:18 CDT 2023 on Orion-login-1.HPC.MsState.Edu
> ---------------------------------------------------
> Checkout:                      *SUCCESS*
> Checkout: with commit hash cb972eed for PR 1644
> Checkout: Completed at Wed May 24 18:08:12 CDT 2023
> Build:                         *SUCCESS*
> Build: Completed at Wed May 24 19:17:21 CDT 2023
> Created experiment:            *SUCCESS*
> Case setup: Completed at Wed May 24 19:17:26 CDT 2023 for experiment C96_atm3DVar
> Experiment C96_atm3DVar Terminated: *FAILED*
> Experiment C96_atm3DVar Terminated with 1 tasks failed at Wed May 24 19:35:14 CDT 2023
> Error logs:
> /work2/noaa/stmp/GFS_CI_ROOT/PR/1644/RUNTESTS/C96_atm3DVar/COMROT/C96_atm3DVar/logs/2021122018/gdasfcst.log
> Automated global-workflow Testing Results:
> ```
> 
> ## Machine: Orion
> Start: Wed May 24 19:39:43 CDT 2023 on Orion-login-1.HPC.MsState.Edu
> Checkout: _SUCCESS_ Checkout: with commit hash [cb972ee](https://github.com/NOAA-EMC/global-workflow/commit/cb972eeded0c53fa34929e482c3a16f2e2396558) for PR 1644 Checkout: Completed at Wed May 24 19:41:33 CDT 2023 Build: _SUCCESS_ Build: Completed at Wed May 24 20:51:27 CDT 2023 Failed to create experiment: _FAIL_ C96_atm3DVar Experiment setup: failed at Wed May 24 20:51:31 CDT 2023 for experiment C96_atm3DVar

Error from the log:
`/work2/noaa/stmp/GFS_CI_ROOT/PR/1644/RUNTESTS/C96_atm3DVar/EXPDIR/C96_atm3DVar/config.base: line 78: obsproc_run_ver: unbound variable`

WalterKolczynski-NOAA commented:
Same problem on Hera: `obsproc_run_ver` not defined.

KateFriedman-NOAA committed:
Add default for obsproc_run_ver for workaround

- Remove default when fcst and efcs jobs return to running
ush/load_fv3gfs_modules.sh and thus sourcing run.ver.

Refs #671

KateFriedman-NOAA committed:
Adjust a couple versions in build.s4.ver

Refs #671

KateFriedman-NOAA commented:
This error stems from the fact that the fcst/efcs jobs don't run `ush/load_fv3gfs_modules.sh`, load their own modules, and thus don't source `run.ver` within `ush/load_fv3gfs_modules.sh`. I added `obsproc_run_ver` into `config.base` yesterday after another review and my own testing overnight on Hera also hit failures in the fcst jobs (same error).

Other than making `fcst.sh` and `efcs.sh` source `run.ver`, a quick workaround for now is to add a default value for `obsproc_run_ver` in `config.base` and remove later. I have tested this with my DEAD fcst/efcs jobs on Hera, it works fine. Committed @ https://github.com/NOAA-EMC/global-workflow/pull/1644/commits/e2eca19944ee5a924627c11475b1aeedf3797f7b.

@WalterKolczynski-NOAA please resubmit the CI testing, thanks!

TerrenceMcGuinness-NOAA commented:
Oh shoot, I need to update an env var to get these to read the Labels from this account.  Should launch shortly.

KateFriedman-NOAA committed:
Update GFS version for COMINsyn on WCOSS2

- Change COMINsyn to use "v16.3" for path on WCOSS2 in
workflow/hosts/wcoss2.yaml.
- Need to update later to use compath.py instead of hardcoding.

Refs #671

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/version-files

KateFriedman-NOAA commented:
The failed job log shows:

```
++ exgdas_atmos_chgres_forenkf.sh[107]: /work2/noaa/stmp/GFS_CI_ROOT/PR/1644/global-workflow/ush/getncdimlen /work2/noaa/stmp/GFS_CI_ROOT/PR/1644/RUNTESTS/C96C48_hybatmDA/COMROT/C96C48_hybatmDA/enkfgdas.20211220/18/mem001/model_data/atmos/history/enkfgdas.t18z.atmf006.nc grid_xt
Traceback (most recent call last):
  File "/work2/noaa/stmp/GFS_CI_ROOT/PR/1644/global-workflow/ush/getncdimlen", line 8, in <module>
    import gsi_utils
  File "/work2/noaa/stmp/GFS_CI_ROOT/PR/1644/global-workflow/ush/gsi_utils.py", line 80
    raise ImportError(f"Unable to import netCDF4 module\n{err}")
                                                              ^
SyntaxError: invalid syntax
+ exgdas_atmos_chgres_forenkf.sh[107]: LONB_ENKF=
+ exgdas_atmos_chgres_forenkf.sh[1]: postamble exgdas_atmos_chgres_forenkf.sh 1685116518 1
```
...but this worked fine in my testing on Orion the other day:
```
++ exgdas_atmos_chgres_forenkf.sh[107]: /work/noaa/global/kfriedma/git/feature-version-files/ush/getncdimlen /work/noaa/stmp/kfriedma/comrot/testverfiles/enkfgdas.20220102/06/mem001/model_data/atmos/history/enkfgdas.t06z.atmf006.nc grid_xt
+ exgdas_atmos_chgres_forenkf.sh[107]: LONB_ENKF=192
```
...although now in an extended cycle it's also failing similarly in the gdasanal and gdaseobs jobs:
```
++ exglobal_atmos_analysis.sh[237]: /work/noaa/global/kfriedma/git/feature-version-files/ush/getncdimlen /work/noaa/stmp/kfriedma/comrot/testverfiles/gdas.20220102/06//model_data/atmos/history/gdas.t06z.atmf006.nc grid_xt
Traceback (most recent call last):
  File "/work/noaa/global/kfriedma/git/feature-version-files/ush/getncdimlen", line 8, in <module>
    import gsi_utils
  File "/work/noaa/global/kfriedma/git/feature-version-files/ush/gsi_utils.py", line 80
    raise ImportError(f"Unable to import netCDF4 module\n{err}")
                                                              ^
SyntaxError: invalid syntax
+ exglobal_atmos_analysis.sh[237]: LONB=
+ exglobal_atmos_analysis.sh[1]: postamble exglobal_atmos_analysis.sh 1685119213 1
```

I notice that the loaded modules are in a different order between my earlier cycle test on Orion and the CI test on Orion, as well as the new cycle in my extended test on Orion. Comparing the loaded modules for the same job against the ones listed in the log for the failed jobs:

Job that worked in an earlier cycle on Tuesday:
```
Orion-login-3[185] /work/noaa/stmp/kfriedma/comrot/testverfiles$ head -20 logs/2022010200/enkfgdasechgres.log
Begin echgres.sh at Tue May 23 15:30:16 UTC 2023
+ echgres.sh[7]: . /work/noaa/global/kfriedma/git/feature-version-files/ush/load_fv3gfs_modules.sh
++ load_fv3gfs_modules.sh[4]: [[ NO == \N\O ]]
++ load_fv3gfs_modules.sh[5]: echo 'Loading modules quietly...'
Loading modules quietly...
++ load_fv3gfs_modules.sh[6]: set +x

Currently Loaded Modules:
  1) gempak/7.5.1     12) wgrib2/2.0.8           23) netcdf/4.7.2
  2) ncl/6.6.2        13) prepobs/1.0.1          24) udunits/2.2.26
  3) jasper/2.0.25    14) fit2obs/1.0.0          25) antlr/2.7.7
  4) zlib/1.2.11      15) hpc/1.2.0              26) nco/4.8.1
  5) png/1.6.35       16) hpc-intel/2018.4       27) wgrib/2.0.8
  6) cdo/1.9.5        17) miniconda3/4.6.14      28) metplus/3.1
  7) prod_util/1.2.2  18) hpc-miniconda3/4.6.14  29) module_base.orion
  8) grib_util/1.2.2  19) ufswm/1.0.0            30) impi/2018.4
  9) g2tmpl/1.10.0    20) met/9.1                31) hpc-impi/2018.4
 10) ncdiag/1.0.0     21) intel/2018.4
 11) crtm/2.4.0       22) hdf5/1.10.5
```
Failed CI test (same module order in failing job in my extended test):
```
Orion-login-3[186] /work/noaa/stmp/kfriedma/comrot/testverfiles$ head -20 /work2/noaa/stmp/GFS_CI_ROOT/PR/1644/RUNTESTS/C96C48_hybatmDA/COMROT/C96C48_hybatmDA/logs/2021122018/enkfgdasechgres.log
Begin echgres.sh at Fri May 26 15:55:15 UTC 2023
+ echgres.sh[7]: . /work2/noaa/stmp/GFS_CI_ROOT/PR/1644/global-workflow/ush/load_fv3gfs_modules.sh
++ load_fv3gfs_modules.sh[4]: [[ NO == \N\O ]]
++ load_fv3gfs_modules.sh[5]: echo 'Loading modules quietly...'
Loading modules quietly...
++ load_fv3gfs_modules.sh[6]: set +x

Currently Loaded Modules:
  1) gempak/7.5.1    10) prod_util/1.2.2  19) hpc-intel/2018.4
  2) ncl/6.6.2       11) grib_util/1.2.2  20) miniconda3/4.6.14
  3) jasper/2.0.25   12) g2tmpl/1.10.0    21) hpc-miniconda3/4.6.14
  4) zlib/1.2.11     13) crtm/2.4.0       22) module_base.orion
  5) png/1.6.35      14) wgrib2/2.0.8     23) impi/2018.4
  6) cdo/1.9.5       15) prepobs/1.0.1    24) hpc-impi/2018.4
  7) udunits/2.2.26  16) fit2obs/1.0.0    25) hdf5/1.10.6
  8) antlr/2.7.7     17) hpc/1.2.0        26) netcdf/4.7.4
  9) nco/4.8.1       18) intel/2018.4     27) ncdiag/1.0.0
```

KateFriedman-NOAA commented:
Extended test on Hera another cycle....jobs that use `getncdimlen` are fine so far and the module order has not changed from prior cycles:
```
-bash-4.2$ head -20 logs/2022010400/enkfgdaseobs.log
Begin eobs.sh at Thu May 25 15:41:21 UTC 2023
+ eobs.sh[7]: . /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-version-files/ush/load_fv3gfs_modules.sh
++ load_fv3gfs_modules.sh[4]: [[ NO == \N\O ]]
++ load_fv3gfs_modules.sh[5]: echo 'Loading modules quietly...'
Loading modules quietly...
++ load_fv3gfs_modules.sh[6]: set +x

Currently Loaded Modules:
  1) hpss/hpss        11) nco/4.9.1        21) intel/18.0.5.274
  2) gempak/7.4.2     12) prod_util/1.2.2  22) hpc-intel/18.0.5.274
  3) ncl/6.6.2        13) grib_util/1.2.2  23) miniconda3/4.6.14
  4) jasper/2.0.25    14) g2tmpl/1.10.0    24) hpc-miniconda3/4.6.14
  5) png/1.6.35       15) ncdiag/1.0.0     25) ufswm/1.0.0
  6) grib_api/1.26.1  16) crtm/2.4.0       26) met/9.1
  7) cdo/1.9.5        17) wgrib2/2.0.8     27) metplus/3.1
  8) R/3.5.0          18) prepobs/1.0.1    28) module_base.hera
  9) hdf5/1.10.6      19) fit2obs/1.0.0    29) impi/2018.0.4
 10) netcdf/4.7.4     20) hpc/1.2.0        30) hpc-impi/2018.0.4

 
-bash-4.2$ head -20 logs/2022010406/enkfgdaseobs.log                                                                                                                       
Begin eobs.sh at Fri May 26 17:00:14 UTC 2023
+ eobs.sh[7]: . /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-version-files/ush/load_fv3gfs_modules.sh
++ load_fv3gfs_modules.sh[4]: [[ NO == \N\O ]]
++ load_fv3gfs_modules.sh[5]: echo 'Loading modules quietly...'
Loading modules quietly...
++ load_fv3gfs_modules.sh[6]: set +x

Currently Loaded Modules:
  1) hpss/hpss        11) nco/4.9.1        21) intel/18.0.5.274
  2) gempak/7.4.2     12) prod_util/1.2.2  22) hpc-intel/18.0.5.274
  3) ncl/6.6.2        13) grib_util/1.2.2  23) miniconda3/4.6.14
  4) jasper/2.0.25    14) g2tmpl/1.10.0    24) hpc-miniconda3/4.6.14
  5) png/1.6.35       15) ncdiag/1.0.0     25) ufswm/1.0.0
  6) grib_api/1.26.1  16) crtm/2.4.0       26) met/9.1
  7) cdo/1.9.5        17) wgrib2/2.0.8     27) metplus/3.1
  8) R/3.5.0          18) prepobs/1.0.1    28) module_base.hera
  9) hdf5/1.10.6      19) fit2obs/1.0.0    29) impi/2018.0.4
 10) netcdf/4.7.4     20) hpc/1.2.0        30) hpc-impi/2018.0.4
```
